truckersmp-cli

```diff
@@ -1,512 +1,770 @@
-        >&2 echo "Failed to download mod files."
-        quit 1
-    fi
-}
-
-update_game() {
+    modfiles = []
+    try:
+        for item in json.JSONDecoder().decode(str(files_json, "ascii"))["Files"]:
+            modfiles.append((item["Md5"], item["FilePath"]))
+        if len(modfiles) == 0:
+            raise Exception("File list is empty")
+    except Exception as e:
+        sys.exit("""Failed to parse files.json: {}
+Please report an issue: {}""".format(e, URL.issueurl))
+
+    # compare existing local files with md5sums
+    # and remember missing/wrong files
+    dlfiles = []
+    for md5, jsonfilepath in modfiles:
+        md5hash = hashlib.md5()
+        modfilepath = os.path.join(args.moddir, jsonfilepath[1:])
+        if not os.path.isfile(modfilepath):
+            dlfiles.append(("/files" + jsonfilepath, modfilepath, md5))
+        else:
+            try:
+                with open(modfilepath, "rb") as f:
+                    while True:
+                        buf = f.read(md5hash.block_size * 4096)
+                        if not buf:
+                            break
+                        md5hash.update(buf)
+                if md5hash.hexdigest() != md5:
+                    dlfiles.append(("/files" + jsonfilepath, modfilepath, md5))
+            except Exception as e:
+                sys.exit("Failed to read {}: {}".format(modfilepath, e))
+    if len(dlfiles) > 0:
+        message_dlfiles = "Files to download:\n"
+        for path, _, _ in dlfiles:
+            message_dlfiles += "  {}\n".format(path)
+        logging.info(message_dlfiles.rstrip())
+    else:
+        logging.debug("No files to download")
+
+    # download missing/wrong files
+    if not download_files(URL.dlurl, dlfiles):
+        if not download_files(URL.dlurlalt, dlfiles):
+            # something went wrong
+            sys.exit("Failed to download mod files.")
+
+
+def update_game():
     # make sure steam is closed before updating
     # it's possible to update with the steam client open but the client looses
     # all connectivity and asks for password and steam guard code after restart
-    while pidof steam > /dev/null 2>&1; do
-        print "found running steam, running pkill -x steam"
-        pkill -x steam
-    sleep 5
-    done
-
-    if ! [ -d "$gamedir" ]; then
-        print "mkdir $gamedir"
-        mkdir -p "$gamedir"
-    fi
-
-    # fetch steamcmd if not in $PATH
-    if ! command -v steamcmd > /dev/null 2>&1; then
-        print "steamcmd not found in \$PATH"
-        if ! [ -f "$steamcmddir/steamcmd.sh" ]; then
-            print "Downloading steamcmd"
-            mkdir -p "$steamcmddir"
-            wget -qO- "$steamcmdurl" | tar xvz -C "$steamcmddir"
-        fi
-
-        steamcmd="$steamcmddir/steamcmd.sh"
-    else
-        print "steamcmd found in \$PATH"
-        steamcmd=$(command -v steamcmd)
-    fi
-    print "Steamcmd: $steamcmd"
+    logging.debug("Closing Steam")
+    subproc.call(["steam", "-shutdown"])
```

Comment:
Just found out that this will start steam if it isn't already open and will just close it again. Maybe a runtime check would prevent this.